### DISCUSSION TITLE
LAMBDA_RUNTIME Errors appearing in logs

### DISCUSSION THREAD

**Discussion by jbednar97** [vercel/community#3248] [Help]
### Summary

Hello, I've recently started noticing 500 errors in "Logs" tab in our Next.js project. Most of the errors look like this:
`…[ERROR] [1689348106912] LAMBDA_RUNTIME Failed to post handler success response. Http response code: 400.
RequestId: 3c73e89b-78b1-40a6-9d6a-db45a571401d Error: Runtime exited with error: exit status 128
Runtime.ExitError`

And I see them from various routes in my api folder.

What I also saw was multiple ones coming from /api/auth/[...auth0], and contacted Auth0 support for this one, but we didn't find out anything related to this problem with them.

I wasn't able to reproduce any of these issues locally, and I have no idea how to get the hang of these issues. The issues can cause our users to not see correct UI, and are pretty crucial.

### Steps to Reproduce

I cannot reproduce it locally unfortunately. But it's basically an Next.js setup with Auth0 SDK

**Comment by amyegan** [comment]
Hey @jbednar97. Response code [`400: MALFORMED_REQUEST_HEADER`](https://vercel.com/docs/errors/application-error-codes#400:-malformed_request_header) happens when a request header contains non-UTF-8 characters. Please make sure your request headers (especially cookie values) are valid UTF-8 strings.

**Reply by jbednar97** [reply]
Hello @amyegan , I'm not really inputting any non-utf8 characters into the headers. I also saw some other errors pretty similar to this one in my logs. 

Error:…RequestId: f9b62ef3-ea04-4f2f-bffb-27deb29c60c1 Error: Runtime exited with error: signal: segmentation fault
Runtime.ExitError

And this all is still attached to the Auth0 logic in the app folder, more exactly as a get request in /api/auth/me

**Reply by jakecollins1** [reply]
@jbednar97, did you ever get to a root cause on this issue? I am seeing the same error sporadically on api endpoints in our sveltekit project

**Reply by jbednar97** [reply]
Hey @jakecollins1  I did not unfortunately. The errors just stopped appearing after some time.

**Reply by jakecollins1** [reply]
Dang, well thanks!

**Reply by saltad-dev** [reply]
Hi, I got same error with jbednar97...

**Comment by jakecollins1** [comment]
I am also seeing this error sporadically in our Sveltekit project when calling an api endpoint.  I am also unable to reproduce locally.
`LAMBDA_RUNTIME Failed to post handler success response. Http response code: 400.
RequestId: 5c8e4597-4cd7-4412-994d-0699d352c42a Error: Runtime exited without providing a reason
Runtime.ExitError`
We are not setting any Cookies in our header, or have anything varying in our header that would cause a sporadic fail

**Comment by shako** [comment]
I have the same error regularly. Also nextjs and nextauth:
```
[ERROR] [1700655121997] LAMBDA_RUNTIME Failed to post handler success response. Http response code: 400.
RequestId: ed73e177-16e9-4379-a5ca-16d12352bd3d Error: Runtime exited with error: exit status 128
Runtime.ExitError
``

**Comment by lev-apeiros** [comment]
Just saw this error today (Next.js 14.0.2):
```
RequestId: 8f832dc4-17e5-47c0-90ac-d5ebd55da93d Error: Runtime exited without providing a reason
Runtime.ExitError
```

**Comment by lcasartelli** [comment]
Same here, up to 6% of total requests are failing with the following message:
```
LAMBDA_RUNTIME Failed to post handler success response. Http response code: 400. RequestId: XXX Error: Runtime exited with error: exit status 128 Runtime.ExitError
```
Looks like it's affecting every route of our Next application and if you try the request that just failed it will succeed immediately after. 
We tried different things like:
- Downgraded Next from 14.x to 13.x
- Added `error.tsx` and `global-error.tsx` in app directory to detect the error (It's not working)
- Added [Log Drains](https://vercel.com/docs/observability/log-drains-overview/log-drains) hopping to get more details from function logs
- Tried vercel CLI locally to emulate Vercel environments

None of these worked in mitigating or getting additional details to solve the issue.

@amyegan can you suggest how to catch meaningful errors from Vercel functions based on these failures?

**Reply by amyegan** [reply]
I've seen intermittent errors like this happen because of an `ECONNRESET` from an endpoint. If the function uses a database or other external service, it's possible the connection is being closed and causing the error.

You may be able to get some connection error info from the service you're connecting to. If there is no attempted connection log available, you could test with a function that does not connect to any external service and see if the error still happens.

https://vercel.com/guides/connection-pooling-with-serverless-functions

**Comment by mcsdevv** [comment]
Hi all, thanks for raising your concerns over the increase in `LAMBDA_RUNTIME` errors you’re experiencing, I’m sharing an update here with the latest.

We’re currently investigating why this is returned in the error log rather than the actual error itself. We suspect it’s a result of unhandled rejections but are working to confirm this.

As of now, our best recommendation is to ensure that every returned promise has an error handler attached to it and we will provide further guidance when available. One thing we can be sure of is that if the status code in the error message is `413`, this means that the request was rejected with a `payload too large` error and in those cases recommend inspecting the function payload size to ensure it’s less than `6MB`.

As further information becomes available, I'll be sure to keep y'all updated here also.

**Comment by belmerbelandres** [comment]
Hello all, We are having exactly the same issue of `Error: Runtime exited with error: exit status 128`.  Anybody have come up for a work around?

**Comment by iwasrobbed** [comment]
Experiencing this too; it rejects the endpoint with a 405 error and then in the logs has `Error: Runtime exited without providing a reason
Runtime.ExitError`

The code where it breaks:

```typescript
const url = '<insert some link to a text file here>'
// aborts on the fetch
const response = await fetch(url)
// axios fails too, not just fetch
// no code after the fetch is reached, including try/catch or console logging
```

- The url doesn't seem to matter. It can be a file in S3, a website, etc.
- Headers don't seem bad / no weird characters
- Wrapping that in a try catch does nothing to catch it either.
- Works fine locally and didn't used to do this (I checked this code a couple weeks ago when I released it and it worked fine). Now it's persistently broken 
- File size doesn't matter; I've had the error on a 4kb or 1mb file. And I have a 300s timeout duration configured

**Reply by belmerbelandres** [reply]
Happens to me with a 500 error and it's random. I'm using canvas too but on very specific page, still happens anywhere on the page. Refreshing the page will work again. This is on production build- can't replicate on local.

**Comment by iwasrobbed** [comment]
Seems like a compiler poisoning issue of some kind; starting happening after installing the `canvas` package `"canvas": "^2.11.2"`

Anytime I'd have a `fetch` within an `api/` route after that, it'd persistently error out.

After removing `canvas` and all the dependencies, it cleared right up

**Comment by mcsdevv** [comment]
Hi all, a further update on this issue which has now been resolved.

We made a change today after identifying a fix that will ensure all stack traces are correctly provided in the logs, rather than the generic `LAMBDA_RUNTIME` error you’ve been seeing.

In some cases, this will have taken effect already, however, the recommended path forward if you are not experiencing an improvement is to redeploy the application.

Thank you all for raising this and your patience in awaiting a resolution :pray:

**Reply by IGassmann** [reply]
Should we open a new "discussion" with the now available stack trace?

**Reply by mcsdevv** [reply]
Hi @IGassmann you're more than welcome to if you're looking for further help!

For clarity, all stack traces now generate will indicate an issue at the code level. While this was always the case, the fact you weren't seeing the actual stack trace and receiving a generic error instead was a problem and has now been resolved.

**Reply by mcsdevv** [reply]
Hi all, we’ve identified an issue with the `@sentry/nextjs package` that relates to this and have been working with the Sentry team on fixing the root cause.

For all using `@sentry/nextjs` please upgrade to the `7.87.0` version as this applies a fix to prevent unhandled errors.

**Reply by charanjit-singh** [reply]
Doesn't work

**Reply by mcsdevv** [reply]
> Doesn't work
> 
> 

Hey @charanjit-singh! Are you able to provide some more details beyond the fact it's not working as expected?

Happy to help!
